app/controllers/users_controller.rb

```diff
@@ -30,4 +65,21 @@ def user_params
                   :password,
                   :password_confirmation)
   end
+
+  def logged_in_user
+    return if logged_in?
+
+    store_location
+    flash[:danger] = t "flash.require_login"
+    redirect_to login_url
+  end
+
+  def correct_user
+    @user = User.find_by id: params[:id]
+    redirect_to(root_url) unless current_user?(@user)
+  end
+
+  def admin_user
+    redirect_to(root_url) unless current_user.admin?
```

Comment:
thêm lý do để thân thiện ng dùng nào
```suggestion
    return if current_user.admin?
    
    flash[:error] = ...
    redirect_to(root_url) 
```